src/client_side_request.h

```diff
@@ -30,17 +30,16 @@ class MemObject;
 /* client_side_request.c - client side request related routines (pure logic) */
 int clientBeginRequest(const HttpRequestMethod&, char const *, CSCB *, CSD *, ClientStreamData, HttpHeader const *, char *, size_t, const MasterXactionPointer &);
 
-class ClientHttpRequest
+class ClientHttpRequest : NonCopyable
```

Comment:
Details of how the trait is implemented are not necessary to expose in the child class public APIs. The effect of removing implicit operator/constructor definitions is sufficient and complete.

Reply:
> Details of how the trait is implemented are not necessary to expose in the child class public APIs.

What "trait"?! In PR code, there is no trait from the child class user point of view! To a (child) class user, private base classes are unimportant -- they are an implementation detail. The class user can and should ignore them. The fact that this (child) class is not copyable is not an implementation detail. It is a part of the class API and should be declared as such from this high-level design point of view.

Also, as detailed in another change request:

* The fact that the (child) class is not to be copied is going to be a part of that (child) class API from the compiler point of view because of the way C++ treats these four special member functions. Whether they are a part of the private or public base has no bearing on this.
* Trying to "hide" that part of the API in a private base gives no advantages and causes problems.

Reply:
The trait of not having implicitly declared copy/move ability. When inheriting from this type definition the child class public API has the trait.

AFAICT, we are saying the same things - but concluding opposite implementation.

IME, we do not need it to be public (same behaviour either way) - so best to leave it as private and remove complexity from objects at the child-of-child-* levels of inheritance.

Reply:
After researching the choice of trait inheritance (public vs private), I am now even more convinced that this change request is correct. I was worried that I am missing something because some popular implementations like Boost::noncopyable recommend private inheritance. I suspect I now know why, despite the fact that "public" is the natural/logical choice for stating that a class is NonCopyable, they require private inheritance: They do it because legacy APIs like Boost::noncopyable could not have a defaulted protected destructor (which appeared in C++11, after those APIs were created). We do not have to follow that limitation. A protected default destructor works well in my basic [tests](https://godbolt.org/z/b1478r4zq). It should be added when addressing this change request.

I cannot find any reason to hide a trait meant for outside class users behind private inheritance. I may still be missing something, of course, but this discussion has not produced the evidence of any flaws in that logic.

> The trait of not having implicitly declared copy/move ability

When that trait is not publicly declared, from an API reader point of view, it does not exist. Yes, the compiler will still produce (sometimes worse) errors, but we are not discussing whether some code will compile. We are discussing the best inheritance choice. When we want to tell others something about a class, all other factor being equal, the best choice is to say it publicly -- we are saying it to _others_.

> we do not need it to be public ... so best to leave it as private

This logic is correct when it comes to private implementation details. It does not apply to traits -- announcements about class traits/capabilities/features/etc. With traits, we want to be public because we are making a statement to the public (that will be using the class).